xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -0,0 +1,692 @@
+        cleanUpRdsWatcher();
+      }
+
+      if (virtualHosts != null) {
+        updateRoutes(virtualHosts, rdsName, getActiveVirtualHost());
+      } else if (changedRdsName) {
+        this.rdsName = rdsName;
+        addWatcher(new RdsWatcher(rdsName));
+        logger.log(XdsLogger.XdsLogLevel.INFO, "Start watching RDS resource {0}", rdsName);
+      }
+
+      setData(update);
+      maybePublishConfig();
+    }
+
+    @Override
+    public void onError(Status error) {
+      super.onError(error);
+      xdsConfigWatcher.onError(toContextString(), error);
+    }
+
+    @Override
+    public void onResourceDoesNotExist(String resourceName) {
+      handleDoesNotExist(resourceName);
+      xdsConfigWatcher.onResourceDoesNotExist(toContextString());
+    }
+
+    private void cleanUpRdsWatcher() {
+      TypeWatchers<?> watchers = resourceWatchers.get(XdsRouteConfigureResource.getInstance());
+      if (watchers == null) {
+        return;
+      }
+      RdsWatcher oldRdsWatcher = (RdsWatcher) watchers.watchers.remove(rdsName);
+      if (oldRdsWatcher != null) {
+        cancelWatcher(oldRdsWatcher);
+      }
+    }
+  }
+
+  private class RdsWatcher extends XdsWatcherBase<RdsUpdate> {
+
+    public RdsWatcher(String resourceName) {
+      super(XdsRouteConfigureResource.getInstance(), resourceName);
+    }
+
+    @Override
+    public void onChanged(RdsUpdate update) {
+      RdsUpdate oldData = hasDataValue() ? getData().getValue() : null;
+      VirtualHost oldVirtualHost =
+          (oldData != null)
+          ? RoutingUtils.findVirtualHostForHostName(oldData.virtualHosts, dataPlaneAuthority)
+          : null;
+      setData(update);
+      updateRoutes(update.virtualHosts, resourceName(), oldVirtualHost);
+      maybePublishConfig();
+    }
+
+    @Override
+    public void onError(Status error) {
+      super.onError(error);
+      xdsConfigWatcher.onError(toContextString(), error);
+    }
+
+    @Override
+    public void onResourceDoesNotExist(String resourceName) {
+      handleDoesNotExist(resourceName);
+      xdsConfigWatcher.onResourceDoesNotExist(toContextString());
+    }
+
+    ImmutableList<String> getCdsNames() {
+      if (!hasDataValue() || getData().getValue().virtualHosts == null) {
+        return ImmutableList.of();
+      }
+
+      return ImmutableList.copyOf(getClusterNamesFromVirtualHost(getActiveVirtualHost()));
+    }
+  }
+
+  private class CdsWatcher extends XdsWatcherBase<XdsClusterResource.CdsUpdate> {
+    Map<String, Integer> parentContexts = new HashMap<>();
+
+    CdsWatcher(String resourceName, String parentContext, int depth) {
+      super(CLUSTER_RESOURCE, resourceName);
+      this.parentContexts.put(parentContext, depth);
+    }
+
+    @Override
+    public void onChanged(XdsClusterResource.CdsUpdate update) {
+      switch (update.clusterType()) {
+        case EDS:
+          setData(update);
+          if (!hasWatcher(ENDPOINT_RESOURCE, update.edsServiceName())) {
```

Comment:
The hasWatcher() method was suspicious; it shouldn't be significant whether there was a watcher already. So I looked here, and it seems to not handle multiple parents. I agree it is unlikely that two different clusters use the same endpoint child, but that's still totally permitted and would be working today.

Reply:
Fixed it so that EdsWatcher keeps a list of its parents.